core/arch/arm/kernel/elf_load_private.h

```diff
@@ -123,4 +148,79 @@ static inline void copy_phdr(struct elf_phdr *phdr,
 			 COPY_PHDR(phdr, ((Elf64_Phdr *)state->phdr + idx)));
 }
 
+#define COPY_SHDR(dst, src) \
+	do { \
+		(dst)->sh_type = (src)->sh_type; \
+		(dst)->sh_addr = (src)->sh_addr; \
+		(dst)->sh_size = (src)->sh_size; \
+	} while (0)
+
+static inline void copy_shdr(struct elf_shdr *shdr,
+			     size_t n,
+			     struct elf_load_state *state)
+{
+	DO_ACTION(state, COPY_SHDR(shdr, ((Elf32_Shdr *)state->shdr + n)),
+			 COPY_SHDR(shdr, ((Elf64_Shdr *)state->shdr + n)));
+}
+
+#define COPY_DYN(dst, src) \
+	do { \
+		(dst)->d_tag = (src)->d_tag; \
+		(dst)->d_un.d_val = (src)->d_un.d_val; \
+	} while (0)
+
+static inline bool copy_dyn(struct elf_dyn *dyn,
+				  size_t n,
+				  struct elf_load_state *state)
+{
+	Elf32_Dyn *lower32 = (Elf32_Dyn *)state->dyn;
+	Elf32_Dyn *dyn32 = (Elf32_Dyn *)state->dyn + n;
+	Elf32_Dyn *upper32 = (Elf32_Dyn *)state->dyn +
+			     state->dyn_size / sizeof(Elf32_Dyn);
+	Elf64_Dyn *lower64 __maybe_unused = (Elf64_Dyn *)state->dyn;
+	Elf64_Dyn *dyn64 __maybe_unused = (Elf64_Dyn *)state->dyn + n;
+	Elf64_Dyn *upper64 __maybe_unused = (Elf64_Dyn *)state->dyn +
+					state->dyn_size / sizeof(Elf64_Dyn);
+	void *p;
+
+	DO_ACTION(state, dyn32 = load_no_speculate_cmp(&dyn32, lower32, upper32,
```

Comment:
I'm validating the pointer against the range. `load_no_speculate_cmp(ptr, low, high, failval, cmpptr)` will return `*ptr` if `cmpptr` is between `low` and `high` and `failval` otherwise. So by passing `&ptr` I obtain the `ptr` value.
See https://developer.arm.com/support/arm-security-updates/speculative-processor-vulnerability/compiler-support-for-mitigations section "More complex cases".

Reply:
We should perhaps have some reference in `lib/libutils/ext/include/speculation_barrier.h` for this. Because it's not easy to find this information. A description of the macros would be nice too since the code and comments are quite hard to reverse engineer. However, that can be done in another PR.